packages/common/src/util.js

```diff
@@ -49,3 +68,99 @@ export function normalizeOauthConfig(configuration) {
 
   return configuration;
 }
+
+function separateUrl(fullUrl) {
+  const urlObject = new URL(fullUrl);
+  const baseUrl = `${urlObject.protocol}//${urlObject.host}`;
+  const path = urlObject.pathname;
+  return { baseUrl, path };
+}
+
+export async function request(method, fullUrl, options = {}) {
+  const { baseUrl, path } = separateUrl(fullUrl);
+  const defaultOptions = {
+    headers: {},
+    query: undefined,
+    timeout: {},
+    tls: {},
+    body: undefined,
+    errors: {
+      401: 'Unauthorised',
+      404: 'Not found',
+    },
+  };
+
+  const { headers, query, body, errors, timeout, tls } = {
+    ...defaultOptions,
+    ...options,
+  };
+
+  if (!headers['Content-Type'] && body) {
+    headers['Content-Type'] = 'application/json';
+  }
+
+  const client = getClient(baseUrl);
+
+  const response = await client.request({
+    path: path,
+    query: query,
+    method: method,
+    headers: headers,
+    errors: errors,
+    body: body ? JSON.stringify(body) : undefined,
+    throwOnError: false,
+  });
+
+  const responseBody = await readResponseBody(response);
+
+  if (response.statusCode >= 400) {
+    const errorMessage = errors[response.statusCode]
+      ? errors[response.statusCode]
+      : `${method} request to ${path} failed with status: ${response.statusCode}`;
+
+    const error = new Error(
+      `Request to ${fullUrl} failed with status: ${response.statusCode}`
+    );
+    error.code = response.statusCode;
+    error.url = fullUrl;
+    error.message = errorMessage;
+    throw error;
+  }
+
+  return {
+    code: response.statusCode,
+    headers: response.headers,
+    data: responseBody,
+  };
+}
+
+async function readResponseBody(response) {
+  const contentType = response.headers['content-type'];
+
+  if (contentType?.includes('application/json')) {
```

Comment:
We need to support some kind of option to force the response to be parsed as json. At the very least we should have a TODO in the code at this stage, but I'd rather see an API proposal and an implementation from you.

Reply:
@josephjclark some thoughts on this 
How about we only support only 3 type of response `json`, `text` and `stream`
The default being `stream`.  Then we determine if `json` or `text` based on `content-type`

Because from an adaptor level if i need the response to be `json` i will pass in `Content-Type: application/json`. 
For response `text()`. We can have an array of Mime-Types then we can check if content-type contains an mime-type in that array. This can be the automagically part 
```
const contentType = getContentTypeFromHeaders()
const mimeTypes = ['application/vnd.ms-excel', 'text/csv', 'text/html' ...]

if(mimeTypes.contains(contentType)) await response.text()
```

Then if i don't specify `Content-Type: json` or i didn't specify `Content-Type` but the response has the mimeType that we support for `text()` we return `stream`

Reply:
Added `parseAs` support